core/src/value/deserialize.rs

```diff
@@ -291,9 +301,146 @@ pub fn parse_datetime(buf: &[u8], dt_utc_offset: FixedOffset) -> Result<DateTime
         .context(InvalidDateTimeZone)
 }
 
+macro_rules! check_range {
+    ($s: expr, $e: expr) => {
+        if $s == $e {
+            RangeIsZero {
+                start: $s.to_string(),
+                end: $e.to_string(),
+            }
+            .fail()
+        } else if $s < $e {
+            Ok((Some($s), Some($e)))
+        } else {
+            RangeInversion {
+                start: $s.to_string(),
+                end: $e.to_string(),
+            }
+            .fail()
+        }
+    };
+}
+
+/**
+ *  Looks for a range separator '-'.
+ *  Returns a tuple of two Option\<NaiveDate\>
+ *  None means no upper or lower range is present
+ */
+pub fn parse_date_range(buf: &[u8]) -> Result<(Option<NaiveDate>, Option<NaiveDate>)> {
+    // minimum length of one valid Date (YYYY) and one '-' separator
+    if buf.len() < 5 {
+        return UnexpectedEndOfElement.fail();
+    }
+
+    if let Some(separator) = buf.iter().position(|e| *e == b'-') {
+        let (start, end) = buf.split_at(separator);
+        let end = &end[1..];
+        match separator {
+            0 => Ok((None, Some(parse_date(end)?.0))),
+            i if i == buf.len() - 1 => Ok((Some(parse_date(start)?.0), None)),
+            _ => {
+                let (s, e) = (parse_date(start)?.0, parse_date(end)?.0);
+                check_range!(s, e)
+            }
+        }
+    } else {
+        NoRangeSeparator.fail()
+    }
+}
+
+/**
+ *  Looks for a range separator '-'.
+ *  Returns a tuple of two Option\<NaiveTime\>
+ *  None means no upper or lower range is present
+ */
+pub fn parse_time_range(buf: &[u8]) -> Result<(Option<NaiveTime>, Option<NaiveTime>)> {
+    // minimum length of one valid Time (HH) and one '-' separator
+    if buf.len() < 3 {
+        return UnexpectedEndOfElement.fail();
+    }
+
+    if let Some(separator) = buf.iter().position(|e| *e == b'-') {
+        let (start, end) = buf.split_at(separator);
+        let end = &end[1..];
+        match separator {
+            0 => Ok((None, Some(parse_time(end)?.0))),
+            i if i == buf.len() - 1 => Ok((Some(parse_time(start)?.0), None)),
+            _ => {
+                let (s, e) = (parse_time(start)?.0, parse_time(end)?.0);
+                check_range!(s, e)
+            }
+        }
+    } else {
+        NoRangeSeparator.fail()
+    }
+}
+
+/**
+ *  Looks for a range separator '-'.
+ *  Returns a tuple of two Option\<DateTime\>
+ *  None means no upper or lower range is present
+ */
+pub fn parse_datetime_range(
+    buf: &[u8],
+    dt_utc_offset: FixedOffset,
+) -> Result<(Option<DateTime<FixedOffset>>, Option<DateTime<FixedOffset>>)> {
```

Comment:
Clippy is suggesting to move this one into a type alias, since it's a pretty long definition.

```rust
type DateTimeRange =  (Option<DateTime<FixedOffset>>, Option<DateTime<FixedOffset>>);
```

And I suppose we could do the same for a date range and a time range.